# Exercise 1 -  Sakila Star Schema & ETL  

All the database tables in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://video.udacity-data.com/topher/2021/August/61120e06_pagila-3nf/pagila-3nf.png)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](https://video.udacity-data.com/topher/2021/August/61120d38_pagila-star/pagila-star.png)

# STEP0: Using ipython-sql

- Load ipython-sql: `%load_ext sql`

- To execute SQL queries you write one of the following atop of your cell: 
    - `%sql`
        - For a one-liner SQL query
        - You can access a python var using `$`    
    - `%%sql`
        - For a multi-line SQL query
        - You can **NOT** access a python var using `$`


- Running a connection string like:
`postgresql://postgres:postgres@db:5432/pagila` connects to the database


# STEP1 : Connect to the local database where Pagila is loaded

##  1.1 Create the pagila db and fill it with data
- Adding `"!"` at the beginning of a jupyter cell runs a command in a shell, i.e. we are not running python code but we are running the `createdb` and `psql` postgresql commmand-line utilities

In [1]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f Data/pagila-schema.sql
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f Data/pagila-data.sql

createdb: database creation failed: ERROR:  database "pagila" already exists
psql:Data/pagila-schema.sql:43: ERROR:  type "mpaa_rating" already exists
psql:Data/pagila-schema.sql:53: ERROR:  type "year" already exists
psql:Data/pagila-schema.sql:70: ERROR:  function "_group_concat" already exists with same argument types
psql:Data/pagila-schema.sql:87: ERROR:  function "film_in_stock" already exists with same argument types
psql:Data/pagila-schema.sql:104: ERROR:  function "film_not_in_stock" already exists with same argument types
psql:Data/pagila-schema.sql:149: ERROR:  function "get_customer_balance" already exists with same argument types
psql:Data/pagila-schema.sql:171: ERROR:  function "inventory_held_by_customer" already exists with same argument types
psql:Data/pagila-schema.sql:208: ERROR:  function "inventory_in_stock" already exists with same argument types
psql:Data/pagila-schema.sql:226: ERROR:  function "last_day" already exists with same argument types
psql:Data/pagila-s

## 1.2 Connect to the newly created db

In [2]:
%load_ext sql

In [3]:
DB_ENDPOINT = "127.0.0.1"
DB = 'pagila'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)


postgresql://student:student@127.0.0.1:5432/pagila


In [4]:
%sql $conn_string

'Connected: student@pagila'

# STEP2 : Explore the  3NF Schema

<img src="./pagila-3nf.png" width="50%"/>

## 2.1 How much? What data sizes are we looking at?

In [5]:
nStores = %sql select count(*) from store;
nFilms = %sql select count(*) from film;
nCustomers = %sql select count(*) from customer;
nRentals = %sql select count(*) from rental;
nPayment = %sql select count(*) from payment;
nStaff = %sql select count(*) from staff;
nCity = %sql select count(*) from city;
nCountry = %sql select count(*) from country;

print("nFilms\t\t=", nFilms[0][0])
print("nCustomers\t=", nCustomers[0][0])
print("nRentals\t=", nRentals[0][0])
print("nPayment\t=", nPayment[0][0])
print("nStaff\t\t=", nStaff[0][0])
print("nStores\t\t=", nStores[0][0])
print("nCities\t\t=", nCity[0][0])
print("nCountry\t\t=", nCountry[0][0])

 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.
nFilms		= 1000
nCustomers	= 599
nRentals	= 16044
nPayment	= 80245
nStaff		= 2
nStores		= 2
nCities		= 600
nCountry		= 109


## 2.2 When? What time period are we talking about?

In [6]:
%%sql 
select min(payment_date) as start, max(payment_date) as end from payment;

 * postgresql://student:***@127.0.0.1:5432/pagila
1 rows affected.


[(datetime.datetime(2017, 1, 24, 21, 21, 56, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), datetime.datetime(2017, 5, 14, 13, 44, 29, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)))]

## 2.3 Where? Where do events in this database occur?
TODO: Write a query that displays the number of addresses by district in the address table. Limit the table to the top 10 districts. Your results should match the table below.

In [7]:
%%sql
select district, sum(city_id) as n
from address
GROUP BY district
ORDER BY n DESC
LIMIT 10;


 * postgresql://student:***@127.0.0.1:5432/pagila
10 rows affected.


[('Shandong', 3237),
 ('England', 2974),
 ('So Paulo', 2952),
 ('West Bengali', 2623),
 ('Buenos Aires', 2572),
 ('Uttar Pradesh', 2462),
 ('California', 2444),
 ('Southern Tagalog', 1931),
 ('Tamil Nadu', 1807),
 ('Hubei', 1790)]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>district</th>
        <th>n</th>
    </tr>
    <tr>
        <td>Shandong</td>
        <td>3237</td>
    </tr>
    <tr>
        <td>England</td>
        <td>2974</td>
    </tr>
    <tr>
        <td>So Paulo</td>
        <td>2952</td>
    </tr>
    <tr>
        <td>West Bengali</td>
        <td>2623</td>
    </tr>
    <tr>
        <td>Buenos Aires</td>
        <td>2572</td>
    </tr>
    <tr>
        <td>Uttar Pradesh</td>
        <td>2462</td>
    </tr>
    <tr>
        <td>California</td>
        <td>2444</td>
    </tr>
    <tr>
        <td>Southern Tagalog</td>
        <td>1931</td>
    </tr>
    <tr>
        <td>Tamil Nadu</td>
        <td>1807</td>
    </tr>
    <tr>
        <td>Hubei</td>
        <td>1790</td>
    </tr>
</tbody></table></div>

# STEP3: Perform some simple data analysis

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed steps 1 and 2, then skip to the second cell.

In [ ]:
# !PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f Data/pagila-schema.sql
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila -f Data/pagila-data.sql

In [13]:
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

The sql extension is already loaded. To reload it, use:
  %reload_ext sql
postgresql://student:student@127.0.0.1:5432/pagila


'Connected: student@pagila'

### 3NF - Entity Relationship Diagram

<img src="./pagila-3nf.png" width="50%"/>

## 3.1 Insight 1:   Top Grossing Movies 
- Payments amounts are in table `payment`
- Movies are in table `film`
- They are not directly linked, `payment` refers to a `rental`, `rental` refers to an `inventory` item and `inventory` item refers to a `film`
- `payment` &rarr; `rental` &rarr; `inventory` &rarr; `film`

### 3.1.1 Films

In [14]:
%%sql
select film_id, title, release_year, rental_rate, rating  from film limit 5;


 * postgresql://student:***@127.0.0.1:5432/pagila
5 rows affected.


[(1, 'ACADEMY DINOSAUR', 2006, Decimal('0.99'), 'PG'),
 (2, 'ACE GOLDFINGER', 2006, Decimal('4.99'), 'G'),
 (3, 'ADAPTATION HOLES', 2006, Decimal('2.99'), 'NC-17'),
 (4, 'AFFAIR PREJUDICE', 2006, Decimal('2.99'), 'G'),
 (5, 'AFRICAN EGG', 2006, Decimal('2.99'), 'G')]

### 3.1.2 Payments

In [15]:
%%sql
select * from payment limit 5;

 * postgresql://student:***@127.0.0.1:5432/pagila
5 rows affected.


[(16050, 269, 2, 7, Decimal('1.99'), datetime.datetime(2017, 1, 24, 21, 40, 19, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (16051, 269, 1, 98, Decimal('0.99'), datetime.datetime(2017, 1, 25, 15, 16, 50, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (16052, 269, 2, 678, Decimal('6.99'), datetime.datetime(2017, 1, 28, 21, 44, 14, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (16053, 269, 2, 703, Decimal('0.99'), datetime.datetime(2017, 1, 29, 0, 58, 2, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (16054, 269, 1, 750, Decimal('4.99'), datetime.datetime(2017, 1, 29, 8, 10, 6, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)))]

### 3.1.3 Inventory

In [16]:
%%sql
select * from inventory limit 5;

 * postgresql://student:***@127.0.0.1:5432/pagila
5 rows affected.


[(1, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (2, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (3, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (4, 1, 1, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))),
 (5, 1, 2, datetime.datetime(2017, 2, 15, 10, 9, 17, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)))]

### 3.1.4 Get the movie of every payment

In [17]:
%%sql
SELECT f.title, p.amount, p.payment_date, p.customer_id                                            
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
limit 5;

 * postgresql://student:***@127.0.0.1:5432/pagila
5 rows affected.


[('SWARM GOLD', Decimal('1.99'), datetime.datetime(2017, 1, 24, 21, 40, 19, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 269),
 ('PACKER MADIGAN', Decimal('0.99'), datetime.datetime(2017, 1, 25, 15, 16, 50, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 269),
 ('SOMETHING DUCK', Decimal('6.99'), datetime.datetime(2017, 1, 28, 21, 44, 14, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 269),
 ('DRACULA CRYSTAL', Decimal('0.99'), datetime.datetime(2017, 1, 29, 0, 58, 2, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 269),
 ('CLOSER BANG', Decimal('4.99'), datetime.datetime(2017, 1, 29, 8, 10, 6, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 269)]

### 3.1.5 sum movie rental revenue
TODO: Write a query that displays the amount of revenue from each title. Limit the results to the top 10 grossing titles. Your results should match the table below.

In [21]:
%%sql
SELECT f.title, sum(amount) revenue                   
FROM film f
JOIN inventory i on f.film_id = i.film_id
JOIN rental r on i.inventory_id = r.inventory_id
JOIN payment p on r.rental_id = p.rental_id
GROUP BY title
ORDER BY revenue DESC
LIMIT 10;

 * postgresql://student:***@127.0.0.1:5432/pagila
10 rows affected.


[('TELEGRAPH VOYAGE', Decimal('1158.65')),
 ('WIFE TURN', Decimal('1118.45')),
 ('ZORRO ARK', Decimal('1073.45')),
 ('GOODFELLAS SALUTE', Decimal('1048.45')),
 ('SATURDAY LAMBS', Decimal('1023.60')),
 ('TITANS JERK', Decimal('1008.55')),
 ('TORQUE BOUND', Decimal('993.60')),
 ('HARRY IDAHO', Decimal('978.50')),
 ('INNOCENT USUAL', Decimal('958.70')),
 ('HUSTLER PARTY', Decimal('953.90'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>title</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>TELEGRAPH VOYAGE</td>
        <td>231.73</td>
    </tr>
    <tr>
        <td>WIFE TURN</td>
        <td>223.69</td>
    </tr>
    <tr>
        <td>ZORRO ARK</td>
        <td>214.69</td>
    </tr>
    <tr>
        <td>GOODFELLAS SALUTE</td>
        <td>209.69</td>
    </tr>
    <tr>
        <td>SATURDAY LAMBS</td>
        <td>204.72</td>
    </tr>
    <tr>
        <td>TITANS JERK</td>
        <td>201.71</td>
    </tr>
    <tr>
        <td>TORQUE BOUND</td>
        <td>198.72</td>
    </tr>
    <tr>
        <td>HARRY IDAHO</td>
        <td>195.70</td>
    </tr>
    <tr>
        <td>INNOCENT USUAL</td>
        <td>191.74</td>
    </tr>
    <tr>
        <td>HUSTLER PARTY</td>
        <td>190.78</td>
    </tr>
</tbody></table></div>

## 3.2 Insight 2:   Top grossing cities 
- Payments amounts are in table `payment`
- Cities are in table `cities`
- `payment` &rarr; `customer` &rarr; `address` &rarr; `city`

### 3.2.1 Get the city of each payment

In [22]:
%%sql
SELECT p.customer_id, p.rental_id, p.amount, ci.city                            
FROM payment p
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
order by p.payment_date
limit 10;

 * postgresql://student:***@127.0.0.1:5432/pagila
10 rows affected.


[(130, 1, Decimal('2.99'), 'guas Lindas de Gois'),
 (130, 1, Decimal('2.99'), 'guas Lindas de Gois'),
 (130, 1, Decimal('2.99'), 'guas Lindas de Gois'),
 (130, 1, Decimal('2.99'), 'guas Lindas de Gois'),
 (130, 1, Decimal('2.99'), 'guas Lindas de Gois'),
 (459, 2, Decimal('2.99'), 'Qomsheh'),
 (459, 2, Decimal('2.99'), 'Qomsheh'),
 (459, 2, Decimal('2.99'), 'Qomsheh'),
 (459, 2, Decimal('2.99'), 'Qomsheh'),
 (459, 2, Decimal('2.99'), 'Qomsheh')]

### 3.2.2 Top grossing cities
TODO: Write a query that returns the total amount of revenue by city as measured by the `amount` variable in the `payment` table. Limit the results to the top 10 cities. Your result should match the table below.

In [23]:
%%sql
SELECT #write code
FROM #write code

 * postgresql://student:***@127.0.0.1:5432/pagila
(psycopg2.ProgrammingError) syntax error at or near "#"
LINE 2: FROM #write code
             ^
 [SQL: 'SELECT #write code\nFROM #write code']


<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>Cape Coral</td>
        <td>221.55</td>
    </tr>
    <tr>
        <td>Saint-Denis</td>
        <td>216.54</td>
    </tr>
    <tr>
        <td>Aurora</td>
        <td>198.50</td>
    </tr>
    <tr>
        <td>Molodetno</td>
        <td>195.58</td>
    </tr>
    <tr>
        <td>Apeldoorn</td>
        <td>194.61</td>
    </tr>
    <tr>
        <td>Santa Brbara dOeste</td>
        <td>194.61</td>
    </tr>
    <tr>
        <td>Qomsheh</td>
        <td>186.62</td>
    </tr>
    <tr>
        <td>London</td>
        <td>180.52</td>
    </tr>
    <tr>
        <td>Ourense (Orense)</td>
        <td>177.60</td>
    </tr>
    <tr>
        <td>Bijapur</td>
        <td>175.61</td>
    </tr>
</tbody></table></div>

## 3.3 Insight 3 : Revenue of a movie by customer city and by month 

### 3.3.1 Total revenue by month

In [24]:
%%sql
SELECT sum(p.amount) as revenue, EXTRACT(month FROM p.payment_date) as month
from payment p
group by month
order by revenue desc
limit 10;

 * postgresql://student:***@127.0.0.1:5432/pagila
5 rows affected.


[(Decimal('142797.30'), 4.0),
 (Decimal('119432.80'), 3.0),
 (Decimal('48159.40'), 2.0),
 (Decimal('24122.15'), 1.0),
 (Decimal('2570.90'), 5.0)]

### 3.3.2 Each movie by customer city and by month (data cube)

In [25]:
%%sql
SELECT f.title, p.amount, p.customer_id, ci.city, p.payment_date,EXTRACT(month FROM p.payment_date) as month
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
order by p.payment_date
limit 10;

 * postgresql://student:***@127.0.0.1:5432/pagila
10 rows affected.


[('BLANKET BEVERLY', Decimal('2.99'), 130, 'guas Lindas de Gois', datetime.datetime(2017, 1, 24, 21, 21, 56, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 1.0),
 ('BLANKET BEVERLY', Decimal('2.99'), 130, 'guas Lindas de Gois', datetime.datetime(2017, 1, 24, 21, 21, 56, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 1.0),
 ('BLANKET BEVERLY', Decimal('2.99'), 130, 'guas Lindas de Gois', datetime.datetime(2017, 1, 24, 21, 21, 56, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 1.0),
 ('BLANKET BEVERLY', Decimal('2.99'), 130, 'guas Lindas de Gois', datetime.datetime(2017, 1, 24, 21, 21, 56, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 1.0),
 ('BLANKET BEVERLY', Decimal('2.99'), 130, 'guas Lindas de Gois', datetime.datetime(2017, 1, 24, 21, 21, 56, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 1.0),
 ('FREAKY POCUS', Decimal('2.99'), 459, 'Qomsheh', datetime.datetime(2017, 1, 24, 21, 22, 59, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 1.0),
 ('FREAKY POCUS', Decimal('2.99'), 459, 'Qomsheh', datetime.datetime(2017, 1, 24, 21, 22, 59, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 1.0),
 ('FREAKY POCUS', Decimal('2.99'), 459, 'Qomsheh', datetime.datetime(2017, 1, 24, 21, 22, 59, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 1.0),
 ('FREAKY POCUS', Decimal('2.99'), 459, 'Qomsheh', datetime.datetime(2017, 1, 24, 21, 22, 59, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 1.0),
 ('FREAKY POCUS', Decimal('2.99'), 459, 'Qomsheh', datetime.datetime(2017, 1, 24, 21, 22, 59, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 1.0)]

### 3.3.3 Sum of revenue of each movie by customer city and by month

TODO: Write a query that returns the total amount of revenue for each movie by customer city and by month. Limit the results to the top 10 movies. Your result should match the table below.

In [31]:
%%sql
SELECT f.title, city, EXTRACT(month FROM p.payment_date) as month, sum(amount) as revenue
FROM film f
JOIN inventory i on f.film_id = i.film_id
JOIN rental r on i.inventory_id = r.inventory_id
JOIN payment p on r.rental_id = p.rental_id
JOIN customer c on r.customer_id = c.customer_id
JOIN store s on c.store_id = s.store_id
JOIN address a on s.address_id = a.address_id
JOIN city ci on a.city_id = ci.city_id
GROUP BY f.title, city, month
ORDER BY month ASC, revenue DESC
LIMIT 10;

 * postgresql://student:***@127.0.0.1:5432/pagila
10 rows affected.


[('SEATTLE EXPECATIONS', 'Lethbridge', 1.0, Decimal('159.80')),
 ('WHALE BIKINI', 'Lethbridge', 1.0, Decimal('104.85')),
 ('ENEMY ODDS', 'Lethbridge', 1.0, Decimal('104.80')),
 ('VIDEOTAPE ARSENIC', 'Woodridge', 1.0, Decimal('99.85')),
 ('INTERVIEW LIAISONS', 'Lethbridge', 1.0, Decimal('89.90')),
 ('APACHE DIVINE', 'Woodridge', 1.0, Decimal('89.85')),
 ('TEEN APOLLO', 'Woodridge', 1.0, Decimal('89.85')),
 ('MODEL FISH', 'Lethbridge', 1.0, Decimal('84.90')),
 ('MILLION ACE', 'Woodridge', 1.0, Decimal('84.90')),
 ('PANKY SUBMARINE', 'Lethbridge', 1.0, Decimal('84.90'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>title</th>
        <th>city</th>
        <th>month</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>SHOW LORD</td>
        <td>Mannheim</td>
        <td>1.0</td>
        <td>11.99</td>
    </tr>
    <tr>
        <td>AMERICAN CIRCUS</td>
        <td>Callao</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>CASUALTIES ENCINO</td>
        <td>Warren</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>TELEGRAPH VOYAGE</td>
        <td>Naala-Porto</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>KISSING DOLLS</td>
        <td>Toulon</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>MILLION ACE</td>
        <td>Bergamo</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>TITANS JERK</td>
        <td>Kimberley</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>DARKO DORADO</td>
        <td>Bhilwara</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>SUNRISE LEAGUE</td>
        <td>Nagareyama</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>MILLION ACE</td>
        <td>Gaziantep</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
</tbody></table></div>